tracing-subscriber/src/filter/directive.rs

```diff
@@ -646,6 +690,116 @@ mod test {
             .collect()
     }
 
+    fn expect_parse(dirs: impl AsRef<str>) -> Vec<Directive> {
+        dirs.as_ref()
+            .split(',')
+            .map(|s| {
+                s.parse()
+                    .unwrap_or_else(|err| panic!("directive '{:?}' should parse: {}", s, err))
+            })
+            .collect()
+    }
+
+    #[test]
+    fn directive_ordering_by_target_len() {
+        // TODO(eliza): it would be nice to have a property-based test for this
+        // instead.
+        let mut dirs = expect_parse(
+            "foo::bar=debug,foo::bar::baz=trace,foo=info,a_really_long_name_with_no_colons=warn",
+        );
+        dirs.sort_unstable();
+
+        let expected = vec![
+            "foo",
+            "foo::bar",
+            "foo::bar::baz",
+            "a_really_long_name_with_no_colons",
+        ];
+        let sorted = dirs
+            .iter()
+            .map(|d| d.target.as_ref().unwrap())
+            .collect::<Vec<_>>();
+
+        assert_eq!(expected, sorted);
+    }
+    #[test]
+    fn directive_ordering_by_span() {
+        // TODO(eliza): it would be nice to have a property-based test for this
+        // instead.
+        let mut dirs = expect_parse("bar[span]=trace,foo=debug,baz::quux=info,a[span]=warn");
+        dirs.sort_unstable();
+
+        let expected = vec!["a", "foo", "bar", "baz::quux"];
```

Comment:
Is it intended that the directive with a span come after the one without? I don't totally follow the motivation here, but that doesn't sound like "most specific first."

Reply:
@Ralith currently, when iterating over the directives for a given `Metadata`, the iterator is reversed:
https://github.com/tokio-rs/tracing/blob/4f8068a5b2b7ca1298733a959687e58a969be374/tracing-subscriber/src/filter/directive.rs#L376-L379

This is kind of an artifact from before we were implementing our own orderings and using string ordering; we could probably remove the `Iter::rev` call and replace it with a `.reverse()` in the PartialOrd impl.

Reply:
2da92161f0ea0b1f91430079572a6cdb592f96c2 changes this so we reverse the `Ordering`s rather than the iterator.

Reply:
Ahh, that explains a lot.